src/jit/morph.cpp

```diff
@@ -18519,6 +18521,280 @@ void Compiler::fgAddFieldSeqForZeroOffset(GenTree* op1, FieldSeqNode* fieldSeq)
+            }
+        }
+
+        PushValue(node);
+
+        return Compiler::WALK_CONTINUE;
+    }
+
+    fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        GenTree* node = *use;
+
+        switch (node->OperGet())
+        {
+            case GT_LCL_VAR:
+                assert(TopValue(0).Node() == node);
+                TopValue(0).Location(node->AsLclVar()->GetLclNum());
+                break;
+
+            case GT_LCL_FLD:
+                assert(TopValue(0).Node() == node);
+                TopValue(0).Location(node->AsLclFld()->GetLclNum(), node->AsLclFld()->gtLclOffs);
+                break;
+
+            case GT_ADDR:
+                assert(TopValue(1).Node() == node);
+                TopValue(1).Address(TopValue(0));
+                PopValue();
+                break;
+
+            case GT_FIELD:
+                if (node->AsField()->gtFldObj != nullptr)
+                {
+                    assert(TopValue(1).Node() == node);
+                    TopValue(1).Field(TopValue(0), node->AsField()->gtFldOffset);
+                    PopValue();
+                }
+                else
+                {
+                    assert(TopValue(0).Node() == node);
+                }
+                break;
+
+            case GT_OBJ:
+            case GT_BLK:
+            case GT_IND:
+                assert(TopValue(1).Node() == node);
+                TopValue(1).Indir(TopValue(0));
+                PopValue();
+                break;
+
+            case GT_CALL:
+                for (; TopValue(0).Node() != node; PopValue())
+                {
+                    if (TopValue(0).IsAddress())
+                    {
+                        bool fieldOnly = (TopValue(0).Node() == node->AsCall()->gtCallObjp);
+                        MarkAddressExposed(TopValue(0), fieldOnly);
+                    }
+                }
+                break;
+
+            default:
+                for (; TopValue(0).Node() != node; PopValue())
+                {
+                    if (TopValue(0).IsAddress())
+                    {
+                        MarkAddressExposed(TopValue(0));
+                    }
+                }
+                break;
+        }
+
+        assert(TopValue(0).Node() == node);
+        return Compiler::WALK_CONTINUE;
+    }
+
+private:
+    void MarkAddressExposed(const Value& val, bool fieldOnly = false)
+    {
+        assert(val.IsAddress());
+
+        unsigned   lclNum = val.GetLclNum();
+        LclVarDsc* varDsc = &m_compiler->lvaTable[lclNum];
+
+        m_compiler->lvaSetVarAddrExposed(lclNum);
+
+        if (varDsc->lvIsStructField)
+        {
+            if (!fieldOnly)
+            {
+                m_compiler->lvaSetVarAddrExposed(varDsc->lvParentLcl);
+            }
+        }
+#ifdef _TARGET_64BIT_
+        else if (!varDsc->lvIsParam && (genActualType(varDsc->TypeGet()) == TYP_INT))
+        {
+            if (Compiler::gtHasCallOnStack(&m_ancestors))
```

Comment:
This is a hack inherited from the old code that wasn't able to figure out how the address is used exactly. The quick needs to be applied only if the address is passed as a call argument. "has call on stack" is obviously an approximation of that. Changing this causes some diffs so I'll keep it like this for now.